File: ios/Classes/banner/AdropBannerManager.swift
import Foundation
import Flutter
import AdropAds

class AdropBannerManager: NSObject, AdropBannerDelegate {
    let messenger: FlutterBinaryMessenger
    var ads: [String: AdropBanner?] = [:]
    var requestIdMap: [AdropBanner: String] = [:]
    var received: [String: Bool] = [:]

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }

    func create(unitId: String, requestId: String) -> AdropBanner {
        let key = keyOf(unitId, requestId)
        if let banner = ads[key] { return banner! }
        let banner = AdropBanner(unitId: unitId)
        banner.delegate = self
        ads[key] = banner
        requestIdMap[banner] = requestId

        return banner
    }

    func load(unitId: String, requestId: String) {
        create(unitId: unitId, requestId: requestId).load()
    }

    func getAd(unitId: String, requestId: String) -> AdropBanner? {
        return ads[keyOf(unitId, requestId)] as? AdropBanner
    }

    func destroy(unitId: String, requestId: String) {
        let key = keyOf(unitId, requestId)

        DispatchQueue.main.async { [weak self] in
            guard let optionalBanner = self?.ads[key], let banner = optionalBanner else {
                return
            }

            self?.ads.removeValue(forKey: key)
            self?.requestIdMap.removeValue(forKey: banner)
        }
    }

    private func keyOf(_ unitId: String, _ requestId: String) -> String {
        return "\(unitId)_\(requestId)"
    }

    private func adropChannel()-> FlutterMethodChannel {
        return FlutterMethodChannel(name: AdropChannel.invokeChannel, binaryMessenger: messenger)
    }

    func onAdClicked(_ banner: AdropAds.AdropBanner) {
        let args: [String: Any] = ["unitId": banner.unitId, "creativeId": banner.creativeId, "requestId": requestIdMap[banner]]
        adropChannel().invokeMethod(AdropMethod.DID_CLICK_AD, arguments: args)
    }

    func onAdFailedToReceive(_ banner: AdropBanner, _ error: AdropErrorCode) {
        let args: [String: Any] = ["unitId": banner.unitId, "error": AdropErrorCodeToString(code: error), "requestId": requestIdMap[banner]]
        adropChannel().invokeMethod(AdropMethod.DID_FAILED_TO_RECEIVE, arguments: args)

    }

    func onAdReceived(_ banner: AdropBanner) {
        let args: [String: Any] = [
            "unitId": banner.unitId,
            "creativeId": banner.creativeId,
            "requestId": requestIdMap[banner],
            "creativeSizeWidth": banner.creativeSize.width,
            "creativeSizeHeight": banner.creativeSize.height
        ]
        adropChannel().invokeMethod(AdropMethod.DID_RECEIVE_AD, arguments: args)
    }
}

File: android/src/main/kotlin/io/adrop/adrop_ads/banner/AdropBannerManager.kt
package io.adrop.adrop_ads.banner


import android.content.Context
import io.adrop.adrop_ads.bridge.AdropChannel
import io.adrop.adrop_ads.bridge.AdropMethod
import io.adrop.ads.banner.AdropBanner
import io.adrop.ads.banner.AdropBannerListener
import io.adrop.ads.model.AdropErrorCode
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel


class AdropBannerManager(
    private val context: Context?,
    messenger: BinaryMessenger
) : AdropBannerListener {

    private val adropChannel = MethodChannel(messenger, AdropChannel.INVOKE_CHANNEL)
    private val ads: MutableMap<String, AdropBanner?> = mutableMapOf()
    private val requestIdMap: MutableMap<AdropBanner, String> = mutableMapOf()

    private fun create(unitId: String, requestId: String): AdropBanner? {
        if (context == null) return null

        val key = keyOf(unitId, requestId)
        return ads[key] ?: let {
            val banner = ads[key] ?: AdropBanner(context, unitId)
            banner.listener = this
            ads[key] = banner
            requestIdMap[banner] = requestId
            banner
        }
    }

    fun load(unitId: String, requestId: String) {
        create(unitId, requestId)?.load()
    }

    fun getAd(unitId: String, requestId: String): AdropBanner? {
        return ads[keyOf(unitId, requestId)]
    }

    fun destroy(unitId: String, requestId: String) {
        val key = keyOf(unitId, requestId)
        ads[key]?.let {
            requestIdMap.remove(it)
            ads.remove(key)
        }
    }

    private fun keyOf(unitId: String, requestId: String): String {
        return "${unitId}_${requestId}"
    }

    override fun onAdClicked(banner: AdropBanner) {
        val unitId = banner.getUnitId()
        ads[unitId] = banner

        val args = mapOf("unitId" to unitId, "creativeId" to banner.creativeId, "requestId" to requestIdMap[banner])
        adropChannel.invokeMethod(AdropMethod.DID_CLICK_AD, args)
    }

    override fun onAdFailedToReceive(banner: AdropBanner, error: AdropErrorCode) {
        val unitId = banner.getUnitId()
        ads[unitId] = banner

        val args = mapOf("unitId" to unitId, "error" to error.name, "requestId" to requestIdMap[banner])
        adropChannel.invokeMethod(AdropMethod.DID_FAIL_TO_RECEIVE_AD, args)
    }

    override fun onAdReceived(banner: AdropBanner) {
        val unitId = banner.getUnitId()
        ads[unitId] = banner

        val args = mapOf(
            "unitId" to unitId,
            "creativeId" to banner.creativeId,
            "requestId" to requestIdMap[banner],
            "creativeSizeWidth" to banner.creativeSize.width,
            "creativeSizeHeight" to banner.creativeSize.height
        )

        adropChannel.invokeMethod(AdropMethod.DID_RECEIVE_AD, args)
    }

    override fun onAdImpression(banner: AdropBanner) {
    }
}
